<h2>Aim: To predict if a person haas AHD or not</h2>

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r"C:\Users\gjrah\Documents\Data Science\Meta_Scifor\Data Sets\Heart.csv")

In [3]:
df

Unnamed: 0  Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  \
0             1   63    1       typical     145   233    1        2    150   
1             2   67    1  asymptomatic     160   286    0        2    108   
2             3   67    1  asymptomatic     120   229    0        2    129   
3             4   37    1    nonanginal     130   250    0        0    187   
4             5   41    0    nontypical     130   204    0        2    172   
..          ...  ...  ...           ...     ...   ...  ...      ...    ...   
298         299   45    1       typical     110   264    0        0    132   
299         300   68    1  asymptomatic     144   193    1        0    141   
300         301   57    1  asymptomatic     130   131    0        0    115   
301         302   57    0    nontypical     130   236    0        2    174   
302         303   38    1    nonanginal     138   175    0        0    173   

     ExAng  Oldpeak  Slope   Ca        Thal  AHD  
0        0      2.3      3  0.0       fixed   No  
1        1      1.5      2  3.0      normal  Yes  
2        1      2.6      2  2.0  reversable  Yes  
3        0      3.5      3  0.0      normal   No  
4        0      1.4      1  0.0      normal   No  
..     ...      ...    ...  ...         ...  ...  
298      0      1.2      2  0.0  reversable  Yes  
299      0      3.4      2  2.0  reversable  Yes  
300      1      1.2      2  1.0  reversable  Yes  
301      0      0.0      2  1.0      normal  Yes  
302      0      0.0      1  NaN      normal   No  

[303 rows x 15 columns]

<h3>Data Pre-Processing</h3>

In [4]:
#Dropping Unamed: 0 since it has not relavance with our analysis
df = df.drop(columns = "Unnamed: 0")

In [5]:
df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal  AHD  
0        2.3      3  0.0       fixed   No  
1        1.5      2  3.0      normal  Yes  
2        2.6      2  2.0  reversable  Yes  
3        3.5      3  0.0      normal   No  
4        1.4      1  0.0      normal   No  
..       ...    ...  ...         ...  ...  
298      1.2      2  0.0  reversable  Yes  
299      3.4      2  2.0  reversable  Yes  
300      1.2      2  1.0  reversable  Yes  
301      0.0      2  1.0      normal  Yes  
302      0.0      1  NaN      normal   No  

[303 rows x 14 columns]

In [6]:
df['ChestPain'] = df['ChestPain'].astype('category')
df['ChestPain'] = df['ChestPain'].cat.codes

In [7]:
df

Age  Sex  ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  \
0     63    1          3     145   233    1        2    150      0      2.3   
1     67    1          0     160   286    0        2    108      1      1.5   
2     67    1          0     120   229    0        2    129      1      2.6   
3     37    1          1     130   250    0        0    187      0      3.5   
4     41    0          2     130   204    0        2    172      0      1.4   
..   ...  ...        ...     ...   ...  ...      ...    ...    ...      ...   
298   45    1          3     110   264    0        0    132      0      1.2   
299   68    1          0     144   193    1        0    141      0      3.4   
300   57    1          0     130   131    0        0    115      1      1.2   
301   57    0          2     130   236    0        2    174      0      0.0   
302   38    1          1     138   175    0        0    173      0      0.0   

     Slope   Ca        Thal  AHD  
0        3  0.0       fixed   No  
1        2  3.0      normal  Yes  
2        2  2.0  reversable  Yes  
3        3  0.0      normal   No  
4        1  0.0      normal   No  
..     ...  ...         ...  ...  
298      2  0.0  reversable  Yes  
299      2  2.0  reversable  Yes  
300      2  1.0  reversable  Yes  
301      2  1.0      normal  Yes  
302      1  NaN      normal   No  

[303 rows x 14 columns]

<b>Categorical Data to Cat Code</b> <br>
asymptomatic = 0 <br>
nonanginal = 1 <br>
nontypical = 2 <br>
typical = 3 

In [8]:
df['Thal'] = df['Thal'].astype("category")
df['Thal'] = df['Thal'].cat.codes

In [9]:
df

Age  Sex  ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  \
0     63    1          3     145   233    1        2    150      0      2.3   
1     67    1          0     160   286    0        2    108      1      1.5   
2     67    1          0     120   229    0        2    129      1      2.6   
3     37    1          1     130   250    0        0    187      0      3.5   
4     41    0          2     130   204    0        2    172      0      1.4   
..   ...  ...        ...     ...   ...  ...      ...    ...    ...      ...   
298   45    1          3     110   264    0        0    132      0      1.2   
299   68    1          0     144   193    1        0    141      0      3.4   
300   57    1          0     130   131    0        0    115      1      1.2   
301   57    0          2     130   236    0        2    174      0      0.0   
302   38    1          1     138   175    0        0    173      0      0.0   

     Slope   Ca  Thal  AHD  
0        3  0.0     0   No  
1        2  3.0     1  Yes  
2        2  2.0     2  Yes  
3        3  0.0     1   No  
4        1  0.0     1   No  
..     ...  ...   ...  ...  
298      2  0.0     2  Yes  
299      2  2.0     2  Yes  
300      2  1.0     2  Yes  
301      2  1.0     1  Yes  
302      1  NaN     1   No  

[303 rows x 14 columns]

<b> Categorical to CatCode </b>
* Fixed = 0
* Normal = 1
* Reversable = 2

In [10]:
df.isnull().sum()

Age          0
Sex          0
ChestPain    0
RestBP       0
Chol         0
Fbs          0
RestECG      0
MaxHR        0
ExAng        0
Oldpeak      0
Slope        0
Ca           4
Thal         0
AHD          0
dtype: int64

* We have 4 <b>null values</b> in Ca

In [11]:
df.dropna()

Age  Sex  ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  \
0     63    1          3     145   233    1        2    150      0      2.3   
1     67    1          0     160   286    0        2    108      1      1.5   
2     67    1          0     120   229    0        2    129      1      2.6   
3     37    1          1     130   250    0        0    187      0      3.5   
4     41    0          2     130   204    0        2    172      0      1.4   
..   ...  ...        ...     ...   ...  ...      ...    ...    ...      ...   
297   57    0          0     140   241    0        0    123      1      0.2   
298   45    1          3     110   264    0        0    132      0      1.2   
299   68    1          0     144   193    1        0    141      0      3.4   
300   57    1          0     130   131    0        0    115      1      1.2   
301   57    0          2     130   236    0        2    174      0      0.0   

     Slope   Ca  Thal  AHD  
0        3  0.0     0   No  
1        2  3.0     1  Yes  
2        2  2.0     2  Yes  
3        3  0.0     1   No  
4        1  0.0     1   No  
..     ...  ...   ...  ...  
297      2  0.0     2  Yes  
298      2  0.0     2  Yes  
299      2  2.0     2  Yes  
300      2  1.0     2  Yes  
301      2  1.0     1  Yes  

[299 rows x 14 columns]

In [12]:
df.isnull().sum()

Age          0
Sex          0
ChestPain    0
RestBP       0
Chol         0
Fbs          0
RestECG      0
MaxHR        0
ExAng        0
Oldpeak      0
Slope        0
Ca           4
Thal         0
AHD          0
dtype: int64

* We have dropped the null values
* It has affected our dataset from 303 rows to 299 rows

<h2>Preparing For Logistic Regression</h2>

In [13]:
X = df.drop(columns="AHD")
X

Age  Sex  ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  \
0     63    1          3     145   233    1        2    150      0      2.3   
1     67    1          0     160   286    0        2    108      1      1.5   
2     67    1          0     120   229    0        2    129      1      2.6   
3     37    1          1     130   250    0        0    187      0      3.5   
4     41    0          2     130   204    0        2    172      0      1.4   
..   ...  ...        ...     ...   ...  ...      ...    ...    ...      ...   
298   45    1          3     110   264    0        0    132      0      1.2   
299   68    1          0     144   193    1        0    141      0      3.4   
300   57    1          0     130   131    0        0    115      1      1.2   
301   57    0          2     130   236    0        2    174      0      0.0   
302   38    1          1     138   175    0        0    173      0      0.0   

     Slope   Ca  Thal  
0        3  0.0     0  
1        2  3.0     1  
2        2  2.0     2  
3        3  0.0     1  
4        1  0.0     1  
..     ...  ...   ...  
298      2  0.0     2  
299      2  2.0     2  
300      2  1.0     2  
301      2  1.0     1  
302      1  NaN     1  

[303 rows x 13 columns]

In [14]:
Y = df['AHD']
Y

0       No
1      Yes
2      Yes
3       No
4       No
      ... 
298    Yes
299    Yes
300    Yes
301    Yes
302     No
Name: AHD, Length: 303, dtype: object

* Train on X parameter to predict Y

<h2> Logistic Regression </h2>

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X,Y,test_size=0.3,random_state=23)

In [17]:
print("X_train \n")
print(X_train)
print("\n")
print("X_test \n")
print(X_test)

X_train 

     Age  Sex  ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  Oldpeak  \
163   58    0          0     100   248    0        2    122      0      1.0   
260   44    0          1     118   242    0        0    149      0      0.3   
124   65    1          3     138   282    1        2    174      0      1.4   
110   61    0          0     145   307    0        2    146      1      1.0   
13    44    1          2     120   263    0        0    173      0      0.0   
..   ...  ...        ...     ...   ...  ...      ...    ...    ...      ...   
237   46    1          0     120   249    0        2    144      0      0.8   
31    60    1          0     117   230    1        0    160      1      1.4   
40    65    0          0     150   225    0        2    114      0      1.0   
230   52    0          1     136   196    0        2    169      0      0.1   
83    68    1          1     180   274    1        2    150      1      1.6   

     Slope   Ca  Thal  
163      2  0.0  

* We can see that X_train has 70% of the dataset while X_test has 30%

In [18]:
from sklearn.preprocessing import StandardScaler

In [19]:
scaler = StandardScaler()

In [20]:
X_train_scaled = scaler.fit_transform(X_train)

In [21]:
X_test_scaled = scaler.transform(X_test)

In [22]:
X_train_scaled

array([[ 0.39538635, -1.42427927, -0.88829631, ...,  0.60745674,
        -0.6831588 , -0.51503427],
       [-1.1366067 , -1.42427927,  0.12417045, ...,  0.60745674,
         0.3985093 , -0.51503427],
       [ 1.16138287,  0.7021095 ,  2.14910398, ...,  0.60745674,
         0.3985093 , -0.51503427],
       ...,
       [ 1.16138287, -1.42427927, -0.88829631, ...,  0.60745674,
         2.56184552,  1.1146264 ],
       [-0.2611821 , -1.42427927,  0.12417045, ...,  0.60745674,
        -0.6831588 , -0.51503427],
       [ 1.48966709,  0.7021095 ,  0.12417045, ...,  0.60745674,
        -0.6831588 ,  1.1146264 ]])

In [23]:
X_test_scaled

array([[ 0.06710212, -1.42427927, -0.88829631, ...,  0.60745674,
        -0.6831588 , -0.51503427],
       [-0.37061018,  0.7021095 ,  0.12417045, ..., -1.00230362,
        -0.6831588 , -0.51503427],
       [ 1.38023902,  0.7021095 , -0.88829631, ...,  0.60745674,
         1.48017741,  1.1146264 ],
       ...,
       [-0.48003825,  0.7021095 , -0.88829631, ...,  0.60745674,
        -0.6831588 ,  1.1146264 ],
       [-0.04232595,  0.7021095 , -0.88829631, ...,  0.60745674,
         0.3985093 ,  1.1146264 ],
       [ 1.05195479,  0.7021095 , -0.88829631, ...,  0.60745674,
         0.3985093 ,  1.1146264 ]])

In [24]:
from sklearn.linear_model import LogisticRegression

In [27]:
log_reg = LogisticRegression(random_state=0).fit(X_train_scaled,y_train)

ValueError: Input X contains NaN.
LogisticRegression does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

<b>Resolving the Null Issue. </b>

In [30]:
X_train_scaled_df = pd.DataFrame(X_train_scaled)

In [32]:
X_train_scaled_df = X_train_scaled_df.dropna()

In [33]:
X_train_scaled = X_train_scaled.to_numpy()

* The above method, didn't resolve our issue, therefore I decided to use Imputer

In [37]:
# Impute missing values
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(strategy='mean')
X_train_scaled_imputed = imputer.fit_transform(X_train_scaled)
X_test_scaled_imputed = imputer.transform(X_test_scaled)


In [41]:
log_reg = LogisticRegression(random_state=0).fit(X_train_scaled_imputed, y_train)

In [42]:
log_reg.predict(X_train_scaled_imputed)

array(['No', 'No', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'No', 'Yes',
       'No', 'Yes', 'Yes', 'Yes', 'No', 'No', 'No', 'No', 'Yes', 'Yes',
       'No', 'No', 'Yes', 'Yes', 'Yes', 'No', 'No', 'Yes', 'No', 'No',
       'No', 'No', 'Yes', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'Yes',
       'No', 'Yes', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'No', 'No',
       'Yes', 'No', 'No', 'Yes', 'Yes', 'No', 'Yes', 'Yes', 'No', 'Yes',
       'Yes', 'Yes', 'Yes', 'No', 'Yes', 'No', 'No', 'Yes', 'Yes', 'Yes',
       'No', 'Yes', 'Yes', 'No', 'No', 'No', 'Yes', 'Yes', 'No', 'No',
       'No', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'Yes', 'No', 'No',
       'Yes', 'No', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'No', 'Yes',
       'No', 'Yes', 'Yes', 'No', 'Yes', 'No', 'Yes', 'No', 'No', 'No',
       'No', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'Yes', 'No', 'Yes',
       'Yes', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No',
       'Yes', 'Yes', 'Yes', 'No', 'Yes', 'No', 'No', 'Yes', 'No', 'No',

In [62]:
#Predicted the accuracy for train attributes

In [43]:
log_reg.score(X_train_scaled_imputed, y_train)

0.8632075471698113

In [63]:
#Predicted the accuracy for test attributes

In [44]:
log_reg.score(X_test_scaled_imputed, y_test)

0.8021978021978022

<b> Imporving the model </b>

* I added the C intercept 
* <b> C = 0.1 </b> gives better results

In [59]:
log_reg1 = LogisticRegression(
    random_state=0,
    C=0.1,
    fit_intercept=True,
).fit(X_train_scaled_imputed,y_train)

In [60]:
log_reg1.score(X_test_scaled_imputed, y_test)

0.8131868131868132